<!DOCTYPE html>
<html >

<head>

  <meta charset="UTF-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <title>Circular Visualization in R</title>
  <meta name="description" content="This book provides a comprehensive overview of implementing circular visualization in R by cirlize package, espeically focusing on visualizaing high dimentional genomic data and revealing complex relationships by Chord diagram.">
  <meta name="generator" content="bookdown 0.3.16 and GitBook 2.6.7">

  <meta property="og:title" content="Circular Visualization in R" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="This book provides a comprehensive overview of implementing circular visualization in R by cirlize package, espeically focusing on visualizaing high dimentional genomic data and revealing complex relationships by Chord diagram." />
  <meta name="github-repo" content="jokergoo/books" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Circular Visualization in R" />
  
  <meta name="twitter:description" content="This book provides a comprehensive overview of implementing circular visualization in R by cirlize package, espeically focusing on visualizaing high dimentional genomic data and revealing complex relationships by Chord diagram." />
  

<meta name="author" content="Zuguang Gu">



  <meta name="viewport" content="width=device-width, initial-scale=1">
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black">
  
  
<link rel="prev" href="position-transformation.html">
<link rel="next" href="revisit-zooming.html">
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />









<style type="text/css">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
code > span.kw { color: #007020; font-weight: bold; } /* Keyword */
code > span.dt { color: #902000; } /* DataType */
code > span.dv { color: #40a070; } /* DecVal */
code > span.bn { color: #40a070; } /* BaseN */
code > span.fl { color: #40a070; } /* Float */
code > span.ch { color: #4070a0; } /* Char */
code > span.st { color: #4070a0; } /* String */
code > span.co { color: #60a0b0; font-style: italic; } /* Comment */
code > span.ot { color: #007020; } /* Other */
code > span.al { color: #ff0000; font-weight: bold; } /* Alert */
code > span.fu { color: #06287e; } /* Function */
code > span.er { color: #ff0000; font-weight: bold; } /* Error */
code > span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
code > span.cn { color: #880000; } /* Constant */
code > span.sc { color: #4070a0; } /* SpecialChar */
code > span.vs { color: #4070a0; } /* VerbatimString */
code > span.ss { color: #bb6688; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { color: #19177c; } /* Variable */
code > span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code > span.op { color: #666666; } /* Operator */
code > span.bu { } /* BuiltIn */
code > span.ex { } /* Extension */
code > span.pp { color: #bc7a00; } /* Preprocessor */
code > span.at { color: #7d9029; } /* Attribute */
code > span.do { color: #ba2121; font-style: italic; } /* Documentation */
code > span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code > span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code > span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">Circular Visualization in R</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>About</a></li>
<li class="part"><span><b>I General Functionality</b></span></li>
<li class="chapter" data-level="1" data-path="circular-layout.html"><a href="circular-layout.html"><i class="fa fa-check"></i><b>1</b> Circular layout</a></li>
<li class="chapter" data-level="2" data-path="graphics.html"><a href="graphics.html"><i class="fa fa-check"></i><b>2</b> Graphics</a></li>
<li class="chapter" data-level="3" data-path="legends.html"><a href="legends.html"><i class="fa fa-check"></i><b>3</b> Legends</a></li>
<li class="chapter" data-level="4" data-path="high-level-plots.html"><a href="high-level-plots.html"><i class="fa fa-check"></i><b>4</b> Implement high-level plots</a></li>
<li class="chapter" data-level="5" data-path="advanced-layout.html"><a href="advanced-layout.html"><i class="fa fa-check"></i><b>5</b> Advanced layout</a></li>
<li class="part"><span><b>II Applications in Genomics</b></span></li>
<li class="chapter" data-level="6" data-path="initialize-genomic-plot.html"><a href="initialize-genomic-plot.html"><i class="fa fa-check"></i><b>6</b> Initialize with genomic data</a></li>
<li class="chapter" data-level="7" data-path="create-plotting-regions-for-genomic-data.html"><a href="create-plotting-regions-for-genomic-data.html"><i class="fa fa-check"></i><b>7</b> Create plotting regions for genomic data</a></li>
<li class="chapter" data-level="8" data-path="modes-of-input.html"><a href="modes-of-input.html"><i class="fa fa-check"></i><b>8</b> modes for <code>circos.genomicTrack()</code></a></li>
<li class="chapter" data-level="9" data-path="position-transformation.html"><a href="position-transformation.html"><i class="fa fa-check"></i><b>9</b> Position transformation</a></li>
<li class="chapter" data-level="10" data-path="high-level-genomic-functions.html"><a href="high-level-genomic-functions.html"><i class="fa fa-check"></i><b>10</b> High-level genomic functions</a><ul>
<li class="chapter" data-level="10.1" data-path="high-level-genomic-functions.html"><a href="high-level-genomic-functions.html#ideograms"><i class="fa fa-check"></i><b>10.1</b> Ideograms</a></li>
<li class="chapter" data-level="10.2" data-path="high-level-genomic-functions.html"><a href="high-level-genomic-functions.html#heatmaps"><i class="fa fa-check"></i><b>10.2</b> Heatmaps</a></li>
<li class="chapter" data-level="10.3" data-path="high-level-genomic-functions.html"><a href="high-level-genomic-functions.html#labels"><i class="fa fa-check"></i><b>10.3</b> Labels</a></li>
<li class="chapter" data-level="10.4" data-path="high-level-genomic-functions.html"><a href="high-level-genomic-functions.html#genomic-density-and-rainfall-plot"><i class="fa fa-check"></i><b>10.4</b> Genomic density and Rainfall plot</a></li>
</ul></li>
<li class="chapter" data-level="11" data-path="revisit-zooming.html"><a href="revisit-zooming.html"><i class="fa fa-check"></i><b>11</b> Revisit zooming</a></li>
<li class="part"><span><b>III Visualize Relations</b></span></li>
<li class="chapter" data-level="12" data-path="we-go-back-to-the-first-track-and-customize-sector-labels.html"><a href="we-go-back-to-the-first-track-and-customize-sector-labels.html"><i class="fa fa-check"></i><b>12</b> we go back to the first track and customize sector labels</a></li>
<li class="chapter" data-level="13" data-path="make-fun-of-the-package.html"><a href="make-fun-of-the-package.html"><i class="fa fa-check"></i><b>13</b> Make fun of the package</a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Circular Visualization in R</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="high-level-genomic-functions" class="section level1">
<h1><span class="header-section-number">Chapter 10</span> High-level genomic functions</h1>
<p>In this chapter, several high-level functions which create tracks are introduced.</p>
<div id="ideograms" class="section level2">
<h2><span class="header-section-number">10.1</span> Ideograms</h2>
<p><code>circos.initializeWithIdeogram()</code> initializes the circular plot and adds ideogram track if the cytoband data is available. Actually, the ideograms are drawn by <code>circos.genomicIdeogram()</code>. <code>circos.genomicIdeogram()</code> creates a thin track of ideograms and can be used anywhere in the circle. By default it adds ideograms for human genome hg19 (Figure <a href="high-level-genomic-functions.html#fig:ideogram">10.1</a>).</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">circos.initializeWithIdeogram</span>(<span class="dt">plotType =</span> <span class="kw">c</span>(<span class="st">&quot;name&quot;</span>, <span class="st">&quot;axis&quot;</span>))
<span class="kw">circos.track</span>(<span class="dt">ylim =</span> <span class="kw">c</span>(<span class="dv">0</span>, <span class="dv">1</span>))
<span class="kw">circos.genomicIdeogram</span>() <span class="co"># put ideogram as the third track</span>
<span class="kw">circos.genomicIdeogram</span>(<span class="dt">track.height =</span> <span class="fl">0.2</span>)</code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:ideogram"></span>
<img src="circlize-book_files/figure-html/ideogram-1.png" alt="Circular ideograms." width="384" />
<p class="caption">
Figure 10.1: Circular ideograms.
</p>
</div>
</div>
<div id="heatmaps" class="section level2">
<h2><span class="header-section-number">10.2</span> Heatmaps</h2>
<p>Matrix which corresponds to genomic regions can be visualized as heatmaps. Heatmaps complete fill the track and there are connection lines connect heatmaps and positions in the genome. `circos.genomicHeamtap() draws connection lines and heatmaps as two tracks and combines them as a integrated track.</p>
<p>Generally, all numeric columns (excluding the first three columns) in the input data frame are used to make the heatmap. Colors can be specfied as a color matrix or a color mapping function generated by <code>colorRamp2()</code>.</p>
<p>The height of the connection line track and the heatmap track can be set by <code>connection_height</code> and <code>heatmap_height</code> arguments. Also parameters for the styles of lines and rectangle borders can be adjusted, please check the help page of <code>circos.genomicHeatmap()</code>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">circos.initializeWithIdeogram</span>()
bed =<span class="st"> </span><span class="kw">generateRandomBed</span>(<span class="dt">nr =</span> <span class="dv">100</span>, <span class="dt">nc =</span> <span class="dv">4</span>)
col_fun =<span class="st"> </span><span class="kw">colorRamp2</span>(<span class="kw">c</span>(<span class="op">-</span><span class="dv">1</span>, <span class="dv">0</span>, <span class="dv">1</span>), <span class="kw">c</span>(<span class="st">&quot;green&quot;</span>, <span class="st">&quot;black&quot;</span>, <span class="st">&quot;red&quot;</span>))
<span class="kw">circos.genomicHeatmap</span>(bed, <span class="dt">col =</span> col_fun, <span class="dt">side =</span> <span class="st">&quot;inside&quot;</span>, <span class="dt">border =</span> <span class="st">&quot;white&quot;</span>)
<span class="kw">circos.clear</span>()</code></pre></div>
<p>In the left figure in Figure <a href="high-level-genomic-functions.html#fig:genomic-heatmap">10.2</a>, the heatmaps are put inside the normal genomic track. Heatmaps are also be put outside the normal genomic track by setting <code>side = &quot;outside&quot;</code> (Figure <a href="high-level-genomic-functions.html#fig:genomic-heatmap">10.2</a>, right).</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">circos.initializeWithIdeogram</span>(<span class="dt">plotType =</span> <span class="ot">NULL</span>)
<span class="kw">circos.genomicHeatmap</span>(bed, <span class="dt">col =</span> col_fun, <span class="dt">side =</span> <span class="st">&quot;outside&quot;</span>,
    <span class="dt">line_col =</span> <span class="kw">as.numeric</span>(<span class="kw">factor</span>(bed[[<span class="dv">1</span>]])))
<span class="kw">circos.genomicIdeogram</span>()
<span class="kw">circos.clear</span>()</code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:genomic-heatmap"></span>
<img src="circlize-book_files/figure-html/genomic-heatmap-1.png" alt="Genomic heamtaps." width="768" />
<p class="caption">
Figure 10.2: Genomic heamtaps.
</p>
</div>
</div>
<div id="labels" class="section level2">
<h2><span class="header-section-number">10.3</span> Labels</h2>
<p><code>circos.genomicLabels()</code> adds text labels for regions specified. Positions of labels are automatically adjusted so that they do not overlap to each other.</p>
<p>Similar as <code>circos.genomicHeatmap()</code>, <code>circos.genomicLabels()</code> also creates two tracks where one for the connection lines and one for the labels. You can set the height of the labels track to be the maximum width of labels by <code>labels_track = max(strwidth(labels))</code>. <code>padding</code> argument controls the gap between two neighbouring labels.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">circos.initializeWithIdeogram</span>()
bed =<span class="st"> </span><span class="kw">generateRandomBed</span>(<span class="dt">nr =</span> <span class="dv">50</span>, <span class="dt">fun =</span> <span class="cf">function</span>(k) <span class="kw">sample</span>(letters, k, <span class="dt">replace =</span> <span class="ot">TRUE</span>))
bed[<span class="dv">1</span>, <span class="dv">4</span>] =<span class="st"> &quot;aaaaa&quot;</span>
<span class="kw">circos.genomicLabels</span>(bed, <span class="dt">labels.column =</span> <span class="dv">4</span>, <span class="dt">side =</span> <span class="st">&quot;inside&quot;</span>)
<span class="kw">circos.clear</span>()</code></pre></div>
<p>Similarlly, labels can be put outside of the normal genomic track (Figure <a href="high-level-genomic-functions.html#fig:genomic-labels">10.3</a> right).</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">circos.initializeWithIdeogram</span>(<span class="dt">plotType =</span> <span class="ot">NULL</span>)
<span class="kw">circos.genomicLabels</span>(bed, <span class="dt">labels.column =</span> <span class="dv">4</span>, <span class="dt">side =</span> <span class="st">&quot;outside&quot;</span>,
    <span class="dt">col =</span> <span class="kw">as.numeric</span>(<span class="kw">factor</span>(bed[[<span class="dv">1</span>]])), <span class="dt">line_col =</span> <span class="kw">as.numeric</span>(<span class="kw">factor</span>(bed[[<span class="dv">1</span>]])))
<span class="kw">circos.genomicIdeogram</span>()
<span class="kw">circos.clear</span>()</code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:genomic-labels"></span>
<img src="circlize-book_files/figure-html/genomic-labels-1.png" alt="Genomic labels." width="768" />
<p class="caption">
Figure 10.3: Genomic labels.
</p>
</div>
</div>
<div id="genomic-density-and-rainfall-plot" class="section level2">
<h2><span class="header-section-number">10.4</span> Genomic density and Rainfall plot</h2>
<p>Rainfall plots are used to visualize the distribution of genomic regions in the genome. Rainfall plots are particularly useful to identify clusters of regions. In the rainfall plot, each dot represents a region. The x-axis corresponds to the genomic coordinate, and the y-axis corresponds to the minimal distance (log10 transformed) of the region to its two neighbouring regions. A cluster of regions will appear as a “rainfall” in the plot.</p>
<p><code>circos.genomicRainfall()</code> calculates neighbouring distance for each region and draw as points on the plot. Since <code>circos.genomicRainfall()</code> generates data on y-direction (<code>log10(distance)</code>), it is actually a high-level function which creates a new track.</p>
<p>The input data can be a single data frame or a list of data frames.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">circos.genoimcRainfall</span>(bed)
<span class="kw">circos.genoimcRainfall</span>(bed_list, <span class="dt">col =</span> <span class="kw">c</span>(<span class="st">&quot;red&quot;</span>, <span class="st">&quot;green&quot;</span>))</code></pre></div>
<p>However, if the amount of regions in a cluster is high, dots will overlap, and direct assessment of the number and density of regions in the cluster will be impossible. To overcome this limitation, additional tracks are added which visualize the genomic density of regions (defined as the fraction of a genomic window that is covered by genomic regions).</p>
<p><code>circos.genomicDensity()</code> calculates how much a genomic window is covered by regions in <code>bed</code>. It is also a high-level function and creates a new track.</p>
<p>The input data can be a single data frame or a list of data frames.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">circos.genomicDensity</span>(bed)
<span class="kw">circos.genomicDensity</span>(bed, <span class="dt">baseline =</span> <span class="dv">0</span>)
<span class="kw">circos.genomicDensity</span>(bed, <span class="dt">window.size =</span> <span class="fl">1e6</span>)
<span class="kw">circos.genomicDensity</span>(bedlist, <span class="dt">col =</span> <span class="kw">c</span>(<span class="st">&quot;#FF000080&quot;</span>, <span class="st">&quot;#0000FF80&quot;</span>))</code></pre></div>
<p>Following example makes a rainfall plot for differentially methylated regions (DMR) and their genomic densities. In the plot, red corresponds to hyper-methylated DMRs (gain of methylation) and blue corresponds to hypo-methylated DMRs (loss of methylation). You may see how the combination of rainfall track and genomic density track helps to get a more precise inference of the distribution of DMRs on genome (Figure <a href="high-level-genomic-functions.html#fig:genomic-rainfall">10.4</a>).</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">load</span>(<span class="kw">system.file</span>(<span class="dt">package =</span> <span class="st">&quot;circlize&quot;</span>, <span class="st">&quot;extdata&quot;</span>, <span class="st">&quot;DMR.RData&quot;</span>))
<span class="kw">par</span>(<span class="dt">mar =</span> <span class="kw">c</span>(<span class="dv">1</span>, <span class="dv">1</span>, <span class="dv">1</span>, <span class="dv">1</span>))
<span class="kw">circos.initializeWithIdeogram</span>(<span class="dt">chromosome.index =</span> <span class="kw">paste0</span>(<span class="st">&quot;chr&quot;</span>, <span class="dv">1</span><span class="op">:</span><span class="dv">22</span>))

bed_list =<span class="st"> </span><span class="kw">list</span>(DMR_hyper, DMR_hypo)
<span class="kw">circos.genomicRainfall</span>(bed_list, <span class="dt">pch =</span> <span class="dv">16</span>, <span class="dt">cex =</span> <span class="fl">0.4</span>, <span class="dt">col =</span> <span class="kw">c</span>(<span class="st">&quot;#FF000080&quot;</span>, <span class="st">&quot;#0000FF80&quot;</span>))
<span class="kw">circos.genomicDensity</span>(DMR_hyper, <span class="dt">col =</span> <span class="kw">c</span>(<span class="st">&quot;#FF000080&quot;</span>), <span class="dt">track.height =</span> <span class="fl">0.1</span>)
<span class="kw">circos.genomicDensity</span>(DMR_hypo, <span class="dt">col =</span> <span class="kw">c</span>(<span class="st">&quot;#0000FF80&quot;</span>), <span class="dt">track.height =</span> <span class="fl">0.1</span>)</code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:genomic-rainfall"></span>
<img src="circlize-book_files/figure-html/genomic-rainfall-1.png" alt="Genomic rainfall plot and densities." width="576" />
<p class="caption">
Figure 10.4: Genomic rainfall plot and densities.
</p>
</div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">circos.clear</span>()</code></pre></div>

</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="position-transformation.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="revisit-zooming.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script>
require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"google": false,
"weibo": false,
"instapper": false,
"vk": false,
"all": ["facebook", "google", "twitter", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/rstudio/bookdown-demo/edit/master/12-high-level-genomic-functions.Rmd",
"text": "Edit"
},
"download": ["circlize-book.pdf", "circlize-book.epub"],
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
