<!DOCTYPE html>
<html >

<head>

  <meta charset="UTF-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <title>Chapter 10 modes for circos.genomicTrack() | Circular Visualization in R</title>
  <meta name="description" content="This book provides a comprehensive overview of implementing circular visualization in R by cirlize package, espeically focusing on visualizaing high dimentional genomic data and revealing complex relationships by Chord diagram.">
  <meta name="generator" content="bookdown  and GitBook 2.6.7">

  <meta property="og:title" content="Chapter 10 modes for circos.genomicTrack() | Circular Visualization in R" />
  <meta property="og:type" content="book" />
  <meta property="og:url" content="https://bookdown.org/jokergoo/circlize-book/book" />
  <meta property="og:image" content="https://bookdown.org/jokergoo/circlize-book/bookimages/circlize_cover.jpg" />
  <meta property="og:description" content="This book provides a comprehensive overview of implementing circular visualization in R by cirlize package, espeically focusing on visualizaing high dimentional genomic data and revealing complex relationships by Chord diagram." />
  <meta name="github-repo" content="jokergoo/circlize_book" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 10 modes for circos.genomicTrack() | Circular Visualization in R" />
  
  <meta name="twitter:description" content="This book provides a comprehensive overview of implementing circular visualization in R by cirlize package, espeically focusing on visualizaing high dimentional genomic data and revealing complex relationships by Chord diagram." />
  <meta name="twitter:image" content="https://bookdown.org/jokergoo/circlize-book/bookimages/circlize_cover.jpg" />

<meta name="author" content="Zuguang Gu">



  <meta name="viewport" content="width=device-width, initial-scale=1">
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black">
  
  
<link rel="prev" href="genomic-plotting-region.html">
<link rel="next" href="high-level-genomic-functions.html">
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />









<style type="text/css">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
code > span.kw { color: #007020; font-weight: bold; } /* Keyword */
code > span.dt { color: #902000; } /* DataType */
code > span.dv { color: #40a070; } /* DecVal */
code > span.bn { color: #40a070; } /* BaseN */
code > span.fl { color: #40a070; } /* Float */
code > span.ch { color: #4070a0; } /* Char */
code > span.st { color: #4070a0; } /* String */
code > span.co { color: #60a0b0; font-style: italic; } /* Comment */
code > span.ot { color: #007020; } /* Other */
code > span.al { color: #ff0000; font-weight: bold; } /* Alert */
code > span.fu { color: #06287e; } /* Function */
code > span.er { color: #ff0000; font-weight: bold; } /* Error */
code > span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
code > span.cn { color: #880000; } /* Constant */
code > span.sc { color: #4070a0; } /* SpecialChar */
code > span.vs { color: #4070a0; } /* VerbatimString */
code > span.ss { color: #bb6688; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { color: #19177c; } /* Variable */
code > span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code > span.op { color: #666666; } /* Operator */
code > span.bu { } /* BuiltIn */
code > span.ex { } /* Extension */
code > span.pp { color: #bc7a00; } /* Preprocessor */
code > span.at { color: #7d9029; } /* Attribute */
code > span.do { color: #ba2121; font-style: italic; } /* Documentation */
code > span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code > span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code > span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">Circular Visualization in R</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>About</a></li>
<li class="part"><span><b>I General Functionality</b></span></li>
<li class="chapter" data-level="1" data-path="introduction.html"><a href="introduction.html"><i class="fa fa-check"></i><b>1</b> Introduction</a><ul>
<li class="chapter" data-level="1.1" data-path="introduction.html"><a href="introduction.html#principle-of-design"><i class="fa fa-check"></i><b>1.1</b> Principle of design</a></li>
<li class="chapter" data-level="1.2" data-path="introduction.html"><a href="introduction.html#a-qiuck-glance"><i class="fa fa-check"></i><b>1.2</b> A quick glance</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="circular-layout.html"><a href="circular-layout.html"><i class="fa fa-check"></i><b>2</b> Circular layout</a><ul>
<li class="chapter" data-level="2.1" data-path="circular-layout.html"><a href="circular-layout.html#coordinate-transformation"><i class="fa fa-check"></i><b>2.1</b> Coordinate transformation</a></li>
<li class="chapter" data-level="2.2" data-path="circular-layout.html"><a href="circular-layout.html#rules-for-making-the-circular-plot"><i class="fa fa-check"></i><b>2.2</b> Rules for making the circular plot</a></li>
<li class="chapter" data-level="2.3" data-path="circular-layout.html"><a href="circular-layout.html#sectors-and-tracks"><i class="fa fa-check"></i><b>2.3</b> Sectors and tracks</a></li>
<li class="chapter" data-level="2.4" data-path="circular-layout.html"><a href="circular-layout.html#graphic-parameters"><i class="fa fa-check"></i><b>2.4</b> Graphic parameters</a></li>
<li class="chapter" data-level="2.5" data-path="circular-layout.html"><a href="circular-layout.html#create-plotting-regions"><i class="fa fa-check"></i><b>2.5</b> Create plotting regions</a></li>
<li class="chapter" data-level="2.6" data-path="circular-layout.html"><a href="circular-layout.html#update-plotting-regions"><i class="fa fa-check"></i><b>2.6</b> Update plotting regions</a></li>
<li class="chapter" data-level="2.7" data-path="circular-layout.html"><a href="circular-layout.html#panel-fun"><i class="fa fa-check"></i><b>2.7</b> <code>panel.fun</code> argument</a></li>
<li class="chapter" data-level="2.8" data-path="circular-layout.html"><a href="circular-layout.html#other-utilities"><i class="fa fa-check"></i><b>2.8</b> Other utilities</a><ul>
<li class="chapter" data-level="2.8.1" data-path="circular-layout.html"><a href="circular-layout.html#circlize_and_reverse_circlize"><i class="fa fa-check"></i><b>2.8.1</b> <code>circlize()</code> and <code>reverse.circlize()</code></a></li>
<li class="chapter" data-level="2.8.2" data-path="circular-layout.html"><a href="circular-layout.html#convert-functions"><i class="fa fa-check"></i><b>2.8.2</b> The convert functions</a></li>
<li class="chapter" data-level="2.8.3" data-path="circular-layout.html"><a href="circular-layout.html#circos-info-and-circos-clear"><i class="fa fa-check"></i><b>2.8.3</b> <code>circos.info()</code> and <code>circos.clear()</code></a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="3" data-path="graphics.html"><a href="graphics.html"><i class="fa fa-check"></i><b>3</b> Graphics</a><ul>
<li class="chapter" data-level="3.1" data-path="graphics.html"><a href="graphics.html#points"><i class="fa fa-check"></i><b>3.1</b> Points</a></li>
<li class="chapter" data-level="3.2" data-path="graphics.html"><a href="graphics.html#lines"><i class="fa fa-check"></i><b>3.2</b> Lines</a></li>
<li class="chapter" data-level="3.3" data-path="graphics.html"><a href="graphics.html#segments"><i class="fa fa-check"></i><b>3.3</b> Segments</a></li>
<li class="chapter" data-level="3.4" data-path="graphics.html"><a href="graphics.html#text"><i class="fa fa-check"></i><b>3.4</b> Text</a></li>
<li class="chapter" data-level="3.5" data-path="graphics.html"><a href="graphics.html#rectangles"><i class="fa fa-check"></i><b>3.5</b> Rectangles and polygons</a></li>
<li class="chapter" data-level="3.6" data-path="graphics.html"><a href="graphics.html#axes"><i class="fa fa-check"></i><b>3.6</b> Axes</a></li>
<li class="chapter" data-level="3.7" data-path="graphics.html"><a href="graphics.html#circular-arrows"><i class="fa fa-check"></i><b>3.7</b> Circular arrows</a></li>
<li class="chapter" data-level="3.8" data-path="graphics.html"><a href="graphics.html#raster-image"><i class="fa fa-check"></i><b>3.8</b> Raster image</a></li>
<li class="chapter" data-level="3.9" data-path="graphics.html"><a href="graphics.html#links"><i class="fa fa-check"></i><b>3.9</b> Links</a></li>
<li class="chapter" data-level="3.10" data-path="graphics.html"><a href="graphics.html#highlight-sectors-and-tracks"><i class="fa fa-check"></i><b>3.10</b> Highlight sectors and tracks</a></li>
<li class="chapter" data-level="3.11" data-path="graphics.html"><a href="graphics.html#work-with-base-graphic-system"><i class="fa fa-check"></i><b>3.11</b> Work together with the base graphic system</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="legends.html"><a href="legends.html"><i class="fa fa-check"></i><b>4</b> Legends</a></li>
<li class="chapter" data-level="5" data-path="high-level-plots.html"><a href="high-level-plots.html"><i class="fa fa-check"></i><b>5</b> Implement high-level circular plots</a><ul>
<li class="chapter" data-level="5.1" data-path="high-level-plots.html"><a href="high-level-plots.html#circular-barplot"><i class="fa fa-check"></i><b>5.1</b> Circular barplots</a></li>
<li class="chapter" data-level="5.2" data-path="high-level-plots.html"><a href="high-level-plots.html#histograms"><i class="fa fa-check"></i><b>5.2</b> Histograms</a></li>
<li class="chapter" data-level="5.3" data-path="high-level-plots.html"><a href="high-level-plots.html#phylogenetic-trees"><i class="fa fa-check"></i><b>5.3</b> Phylogenetic trees</a></li>
<li class="chapter" data-level="5.4" data-path="high-level-plots.html"><a href="high-level-plots.html#heatmaps"><i class="fa fa-check"></i><b>5.4</b> Heatmaps</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="advanced-layout.html"><a href="advanced-layout.html"><i class="fa fa-check"></i><b>6</b> Advanced layout</a><ul>
<li class="chapter" data-level="6.1" data-path="advanced-layout.html"><a href="advanced-layout.html#zooming-of-sectors"><i class="fa fa-check"></i><b>6.1</b> Zooming of sectors</a></li>
<li class="chapter" data-level="6.2" data-path="advanced-layout.html"><a href="advanced-layout.html#part-circle"><i class="fa fa-check"></i><b>6.2</b> Visualize part of the circle</a></li>
<li class="chapter" data-level="6.3" data-path="advanced-layout.html"><a href="advanced-layout.html#combine-circular-plots"><i class="fa fa-check"></i><b>6.3</b> Combine multiple circular plots</a></li>
<li class="chapter" data-level="6.4" data-path="advanced-layout.html"><a href="advanced-layout.html#arrange-multiple-plots"><i class="fa fa-check"></i><b>6.4</b> Arrange multiple plots</a></li>
</ul></li>
<li class="part"><span><b>II Applications in Genomics</b></span></li>
<li class="chapter" data-level="7" data-path="genomic-introduction.html"><a href="genomic-introduction.html"><i class="fa fa-check"></i><b>7</b> Introduction</a><ul>
<li class="chapter" data-level="7.1" data-path="genomic-introduction.html"><a href="genomic-introduction.html#input-data"><i class="fa fa-check"></i><b>7.1</b> Input data</a></li>
</ul></li>
<li class="chapter" data-level="8" data-path="initialize-genomic-plot.html"><a href="initialize-genomic-plot.html"><i class="fa fa-check"></i><b>8</b> Initialize with genomic data</a><ul>
<li class="chapter" data-level="8.1" data-path="initialize-genomic-plot.html"><a href="initialize-genomic-plot.html#initialize-cytoband"><i class="fa fa-check"></i><b>8.1</b> Initialize with cytoband data</a><ul>
<li class="chapter" data-level="8.1.1" data-path="initialize-genomic-plot.html"><a href="initialize-genomic-plot.html#basic-usage"><i class="fa fa-check"></i><b>8.1.1</b> Basic usage</a></li>
<li class="chapter" data-level="8.1.2" data-path="initialize-genomic-plot.html"><a href="initialize-genomic-plot.html#pre-defined-tracks"><i class="fa fa-check"></i><b>8.1.2</b> Pre-defined tracks</a></li>
<li class="chapter" data-level="8.1.3" data-path="initialize-genomic-plot.html"><a href="initialize-genomic-plot.html#other-general-settings"><i class="fa fa-check"></i><b>8.1.3</b> Other general settings</a></li>
</ul></li>
<li class="chapter" data-level="8.2" data-path="initialize-genomic-plot.html"><a href="initialize-genomic-plot.html#customize-chromosome-track"><i class="fa fa-check"></i><b>8.2</b> Customize chromosome track</a></li>
<li class="chapter" data-level="8.3" data-path="initialize-genomic-plot.html"><a href="initialize-genomic-plot.html#initialize-with-general-genomic-category"><i class="fa fa-check"></i><b>8.3</b> Initialize with general genomic category</a></li>
<li class="chapter" data-level="8.4" data-path="initialize-genomic-plot.html"><a href="initialize-genomic-plot.html#zooming-chromosomes"><i class="fa fa-check"></i><b>8.4</b> Zooming chromosomes</a></li>
</ul></li>
<li class="chapter" data-level="9" data-path="genomic-plotting-region.html"><a href="genomic-plotting-region.html"><i class="fa fa-check"></i><b>9</b> Create plotting regions for genomic data</a><ul>
<li class="chapter" data-level="9.1" data-path="genomic-plotting-region.html"><a href="genomic-plotting-region.html#genomic-points"><i class="fa fa-check"></i><b>9.1</b> Points</a></li>
<li class="chapter" data-level="9.2" data-path="genomic-plotting-region.html"><a href="genomic-plotting-region.html#genomic-lines"><i class="fa fa-check"></i><b>9.2</b> Lines</a></li>
<li class="chapter" data-level="9.3" data-path="genomic-plotting-region.html"><a href="genomic-plotting-region.html#genomic-text"><i class="fa fa-check"></i><b>9.3</b> Text</a></li>
<li class="chapter" data-level="9.4" data-path="genomic-plotting-region.html"><a href="genomic-plotting-region.html#genomic-rectangles"><i class="fa fa-check"></i><b>9.4</b> Rectangles</a></li>
<li class="chapter" data-level="9.5" data-path="genomic-plotting-region.html"><a href="genomic-plotting-region.html#genomic-links"><i class="fa fa-check"></i><b>9.5</b> Links</a></li>
<li class="chapter" data-level="9.6" data-path="genomic-plotting-region.html"><a href="genomic-plotting-region.html#mixed-use-of-general-circlize-functions"><i class="fa fa-check"></i><b>9.6</b> Mixed use of general circlize functions</a></li>
</ul></li>
<li class="chapter" data-level="10" data-path="modes-of-input.html"><a href="modes-of-input.html"><i class="fa fa-check"></i><b>10</b> modes for <code>circos.genomicTrack()</code></a><ul>
<li class="chapter" data-level="10.1" data-path="modes-of-input.html"><a href="modes-of-input.html#normal-mode"><i class="fa fa-check"></i><b>10.1</b> Normal mode</a><ul>
<li class="chapter" data-level="10.1.1" data-path="modes-of-input.html"><a href="modes-of-input.html#input-is-a-data-frame"><i class="fa fa-check"></i><b>10.1.1</b> Input is a data frame</a></li>
<li class="chapter" data-level="10.1.2" data-path="modes-of-input.html"><a href="modes-of-input.html#input-is-a-list-of-data-frames"><i class="fa fa-check"></i><b>10.1.2</b> Input is a list of data frames</a></li>
</ul></li>
<li class="chapter" data-level="10.2" data-path="modes-of-input.html"><a href="modes-of-input.html#stack-mode"><i class="fa fa-check"></i><b>10.2</b> Stack mode</a><ul>
<li class="chapter" data-level="10.2.1" data-path="modes-of-input.html"><a href="modes-of-input.html#input-is-a-data-frame-1"><i class="fa fa-check"></i><b>10.2.1</b> Input is a data frame</a></li>
<li class="chapter" data-level="10.2.2" data-path="modes-of-input.html"><a href="modes-of-input.html#input-is-a-list-of-data-frames-1"><i class="fa fa-check"></i><b>10.2.2</b> Input is a list of data frames</a></li>
</ul></li>
<li class="chapter" data-level="10.3" data-path="modes-of-input.html"><a href="modes-of-input.html#applications"><i class="fa fa-check"></i><b>10.3</b> Applications</a><ul>
<li class="chapter" data-level="10.3.1" data-path="modes-of-input.html"><a href="modes-of-input.html#modes-points"><i class="fa fa-check"></i><b>10.3.1</b> Points</a></li>
<li class="chapter" data-level="10.3.2" data-path="modes-of-input.html"><a href="modes-of-input.html#modes-lines"><i class="fa fa-check"></i><b>10.3.2</b> Lines</a></li>
<li class="chapter" data-level="10.3.3" data-path="modes-of-input.html"><a href="modes-of-input.html#modes-rectangles"><i class="fa fa-check"></i><b>10.3.3</b> Rectangles</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="11" data-path="high-level-genomic-functions.html"><a href="high-level-genomic-functions.html"><i class="fa fa-check"></i><b>11</b> High-level genomic functions</a><ul>
<li class="chapter" data-level="11.1" data-path="high-level-genomic-functions.html"><a href="high-level-genomic-functions.html#ideograms"><i class="fa fa-check"></i><b>11.1</b> Ideograms</a></li>
<li class="chapter" data-level="11.2" data-path="high-level-genomic-functions.html"><a href="high-level-genomic-functions.html#genomic-heatmap"><i class="fa fa-check"></i><b>11.2</b> Heatmaps</a></li>
<li class="chapter" data-level="11.3" data-path="high-level-genomic-functions.html"><a href="high-level-genomic-functions.html#labels"><i class="fa fa-check"></i><b>11.3</b> Labels</a></li>
<li class="chapter" data-level="11.4" data-path="high-level-genomic-functions.html"><a href="high-level-genomic-functions.html#genomic-axes"><i class="fa fa-check"></i><b>11.4</b> Genomic axes</a></li>
<li class="chapter" data-level="11.5" data-path="high-level-genomic-functions.html"><a href="high-level-genomic-functions.html#genomic-density-and-rainfall-plot"><i class="fa fa-check"></i><b>11.5</b> Genomic density and Rainfall plot</a></li>
</ul></li>
<li class="chapter" data-level="12" data-path="nested-zooming.html"><a href="nested-zooming.html"><i class="fa fa-check"></i><b>12</b> Nested zooming</a><ul>
<li class="chapter" data-level="12.1" data-path="nested-zooming.html"><a href="nested-zooming.html#basic-idea"><i class="fa fa-check"></i><b>12.1</b> Basic idea</a></li>
<li class="chapter" data-level="12.2" data-path="nested-zooming.html"><a href="nested-zooming.html#visualization-of-dmrs-from-tagmentation-based-wgbs"><i class="fa fa-check"></i><b>12.2</b> Visualization of DMRs from tagmentation-based WGBS</a></li>
</ul></li>
<li class="part"><span><b>III Visualize Relations</b></span></li>
<li class="chapter" data-level="13" data-path="the-chorddiagram-function.html"><a href="the-chorddiagram-function.html"><i class="fa fa-check"></i><b>13</b> The <code>chordDiagram()</code> function</a><ul>
<li class="chapter" data-level="13.1" data-path="the-chorddiagram-function.html"><a href="the-chorddiagram-function.html#basic-usage-1"><i class="fa fa-check"></i><b>13.1</b> Basic usage</a></li>
<li class="chapter" data-level="13.2" data-path="the-chorddiagram-function.html"><a href="the-chorddiagram-function.html#adjust-by-circos.par"><i class="fa fa-check"></i><b>13.2</b> Adjust by <code>circos.par()</code></a></li>
<li class="chapter" data-level="13.3" data-path="the-chorddiagram-function.html"><a href="the-chorddiagram-function.html#chord-diagram-colors"><i class="fa fa-check"></i><b>13.3</b> Colors</a><ul>
<li class="chapter" data-level="13.3.1" data-path="the-chorddiagram-function.html"><a href="the-chorddiagram-function.html#set-grid-colors"><i class="fa fa-check"></i><b>13.3.1</b> Set grid colors</a></li>
<li class="chapter" data-level="13.3.2" data-path="the-chorddiagram-function.html"><a href="the-chorddiagram-function.html#chord-diagram-link-color"><i class="fa fa-check"></i><b>13.3.2</b> Set link colors</a></li>
</ul></li>
<li class="chapter" data-level="13.4" data-path="the-chorddiagram-function.html"><a href="the-chorddiagram-function.html#chord-diagram-link-border"><i class="fa fa-check"></i><b>13.4</b> Link border</a></li>
<li class="chapter" data-level="13.5" data-path="the-chorddiagram-function.html"><a href="the-chorddiagram-function.html#highlight-links"><i class="fa fa-check"></i><b>13.5</b> Highlight links</a></li>
<li class="chapter" data-level="13.6" data-path="the-chorddiagram-function.html"><a href="the-chorddiagram-function.html#orders-of-links"><i class="fa fa-check"></i><b>13.6</b> Orders of links</a><ul>
<li class="chapter" data-level="13.6.1" data-path="the-chorddiagram-function.html"><a href="the-chorddiagram-function.html#orders-of-positions-on-sectors"><i class="fa fa-check"></i><b>13.6.1</b> Orders of positions on sectors</a></li>
<li class="chapter" data-level="13.6.2" data-path="the-chorddiagram-function.html"><a href="the-chorddiagram-function.html#order-of-adding-links"><i class="fa fa-check"></i><b>13.6.2</b> Order of adding links</a></li>
</ul></li>
<li class="chapter" data-level="13.7" data-path="the-chorddiagram-function.html"><a href="the-chorddiagram-function.html#self-links"><i class="fa fa-check"></i><b>13.7</b> Self-links</a></li>
<li class="chapter" data-level="13.8" data-path="the-chorddiagram-function.html"><a href="the-chorddiagram-function.html#symmetric-matrix"><i class="fa fa-check"></i><b>13.8</b> Symmetric matrix</a></li>
<li class="chapter" data-level="13.9" data-path="the-chorddiagram-function.html"><a href="the-chorddiagram-function.html#directional-relations"><i class="fa fa-check"></i><b>13.9</b> Directional relations</a></li>
<li class="chapter" data-level="13.10" data-path="the-chorddiagram-function.html"><a href="the-chorddiagram-function.html#scaling"><i class="fa fa-check"></i><b>13.10</b> Scaling</a></li>
<li class="chapter" data-level="13.11" data-path="the-chorddiagram-function.html"><a href="the-chorddiagram-function.html#reduce"><i class="fa fa-check"></i><b>13.11</b> Reduce</a></li>
<li class="chapter" data-level="13.12" data-path="the-chorddiagram-function.html"><a href="the-chorddiagram-function.html#input-as-a-data-frame"><i class="fa fa-check"></i><b>13.12</b> Input as a data frame</a><ul>
<li class="chapter" data-level="13.12.1" data-path="the-chorddiagram-function.html"><a href="the-chorddiagram-function.html#multiple-links"><i class="fa fa-check"></i><b>13.12.1</b> Multiple links</a></li>
<li class="chapter" data-level="13.12.2" data-path="the-chorddiagram-function.html"><a href="the-chorddiagram-function.html#unequal-width-of-the-link-ends"><i class="fa fa-check"></i><b>13.12.2</b> Unequal width of the link ends</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="14" data-path="advanced-usage-of-chorddiagram.html"><a href="advanced-usage-of-chorddiagram.html"><i class="fa fa-check"></i><b>14</b> Advanced usage of <code>chordDiagram()</code></a><ul>
<li class="chapter" data-level="14.1" data-path="advanced-usage-of-chorddiagram.html"><a href="advanced-usage-of-chorddiagram.html#organization-of-tracks"><i class="fa fa-check"></i><b>14.1</b> Organization of tracks</a></li>
<li class="chapter" data-level="14.2" data-path="advanced-usage-of-chorddiagram.html"><a href="advanced-usage-of-chorddiagram.html#customize-sector-labels"><i class="fa fa-check"></i><b>14.2</b> Customize sector labels</a></li>
<li class="chapter" data-level="14.3" data-path="advanced-usage-of-chorddiagram.html"><a href="advanced-usage-of-chorddiagram.html#customize-sector-axes"><i class="fa fa-check"></i><b>14.3</b> Customize sector axes</a></li>
<li class="chapter" data-level="14.4" data-path="advanced-usage-of-chorddiagram.html"><a href="advanced-usage-of-chorddiagram.html#put-horizontally-or-vertically-symmetric"><i class="fa fa-check"></i><b>14.4</b> Put horizontally or vertically symmetric</a></li>
<li class="chapter" data-level="14.5" data-path="advanced-usage-of-chorddiagram.html"><a href="advanced-usage-of-chorddiagram.html#compare-two-chord-diagrams"><i class="fa fa-check"></i><b>14.5</b> Compare two Chord diagrams</a></li>
<li class="chapter" data-level="14.6" data-path="advanced-usage-of-chorddiagram.html"><a href="advanced-usage-of-chorddiagram.html#multiple-group-chord-diagram"><i class="fa fa-check"></i><b>14.6</b> Multiple-group Chord diagram</a></li>
</ul></li>
<li class="chapter" data-level="15" data-path="a-complex-example-of-chord-diagram.html"><a href="a-complex-example-of-chord-diagram.html"><i class="fa fa-check"></i><b>15</b> A complex example of Chord diagram</a></li>
<li class="part"><span><b>IV Others</b></span></li>
<li class="chapter" data-level="16" data-path="make-fun-of-the-package.html"><a href="make-fun-of-the-package.html"><i class="fa fa-check"></i><b>16</b> Make fun of the package</a><ul>
<li class="chapter" data-level="16.1" data-path="make-fun-of-the-package.html"><a href="make-fun-of-the-package.html#a-clock"><i class="fa fa-check"></i><b>16.1</b> A clock</a></li>
<li class="chapter" data-level="16.2" data-path="make-fun-of-the-package.html"><a href="make-fun-of-the-package.html#a-dartboard"><i class="fa fa-check"></i><b>16.2</b> A dartboard</a></li>
<li class="chapter" data-level="16.3" data-path="make-fun-of-the-package.html"><a href="make-fun-of-the-package.html#ba-gua-and-tai-ji"><i class="fa fa-check"></i><b>16.3</b> Ba-Gua and Tai-Ji</a></li>
</ul></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Circular Visualization in R</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="modes-of-input" class="section level1">
<h1><span class="header-section-number">Chapter 10</span> modes for <code>circos.genomicTrack()</code></h1>
<p>The behaviour of <code>circos.genomicTrack()</code> and <code>panel.fun</code> will be different according to different input data (e.g. is it a simple data frame or a list of data frames? If it is a data frame, how many numeric columns it has?) and different settings.</p>
<div id="normal-mode" class="section level2">
<h2><span class="header-section-number">10.1</span> Normal mode</h2>
<div id="input-is-a-data-frame" class="section level3">
<h3><span class="header-section-number">10.1.1</span> Input is a data frame</h3>
<p>If input <code>data</code> is a data frame in <em>BED</em> format, <code>region</code> in <code>panel.fun</code> would be a data frame containing start position and end position in the current chromosome which is extracted from <code>data</code>. <code>value</code> is also a data frame which contains columns in <code>data</code> excluding the first three columns. Index of proper numeric columns will be passed by <code>...</code> if it is set in <code>circos.genomicTrack()</code>. If users want to use such information, they need to pass <code>...</code> to low-level genomic function such as <code>circos.genoimcPoints()</code> as well.</p>
<p>If there are more than one numeric columns, graphics are added for each column repeatedly (with same genomic positions).</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">data =<span class="st"> </span><span class="kw">generateRandomBed</span>(<span class="dt">nc =</span> <span class="dv">2</span>)
<span class="kw">circos.genomicTrack</span>(data, <span class="dt">numeric.column =</span> <span class="dv">4</span>, 
    <span class="dt">panel.fun =</span> <span class="cf">function</span>(region, value, ...) {
        <span class="kw">circos.genomicPoints</span>(region, value, ...)
        <span class="kw">circos.genomicPoints</span>(region, value)
        <span class="co"># 1st column in `value` while 4th column in `data`</span>
        <span class="kw">circos.genomicPoints</span>(region, value, <span class="dt">numeric.column =</span> <span class="dv">1</span>)
})</code></pre></div>
</div>
<div id="input-is-a-list-of-data-frames" class="section level3">
<h3><span class="header-section-number">10.1.2</span> Input is a list of data frames</h3>
<p>If input data is a list of data frames, <code>panel.fun</code> is applied on each data frame iteratively to the current cell. Under such condition, <code>region</code> and <code>value</code> will contain corresponding data in the current data frame and in the current chromosome. The index for the current data frame can be get by <code>getI(...)</code>. Note <code>getI(...)</code> can only be used inside <code>panel.fun</code> and <code>...</code> argument is mandatory.</p>
<p>When <code>numeric.column</code> is specified in <code>circos.genomicTrack()</code>, the length of <code>numeric.column</code> can only be one or the number of data frames, which means, there is only one numeric column that will be used in each data frame. If it is not specified, the first numeric column in each data frame is used.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">bed_list =<span class="st"> </span><span class="kw">list</span>(<span class="kw">generateRandomBed</span>(), <span class="kw">generateRandomBed</span>())
<span class="kw">circos.genomicTrack</span>(bed_list,
    <span class="dt">panel.fun =</span> <span class="cf">function</span>(region, value, ...) {
        i =<span class="st"> </span><span class="kw">getI</span>(...)
        <span class="kw">circos.genomicPoints</span>(region, value, <span class="dt">col =</span> i, ...)
})

<span class="co"># column 4 in the first bed and column 5 in the second bed</span>
<span class="kw">circos.genomicTrack</span>(bed_list, <span class="dt">numeric.column =</span> <span class="kw">c</span>(<span class="dv">4</span>, <span class="dv">5</span>),
    <span class="dt">panel.fun =</span> <span class="cf">function</span>(region, value, ...) {
        i =<span class="st"> </span><span class="kw">getI</span>(...)
        <span class="kw">circos.genomicPoints</span>(region, value, <span class="dt">col =</span> i, ...)
})</code></pre></div>
</div>
</div>
<div id="stack-mode" class="section level2">
<h2><span class="header-section-number">10.2</span> Stack mode</h2>
<p><code>circos.genomicTrack()</code> also supports a <code>stack</code> mode by setting <code>stack = TRUE</code>. Under <code>stack</code> mode, <code>ylim</code> is re-defined inside the function and the y-axis is splitted into several bins with equal height and graphics are put onto “horizontal” bins (with position <code>y = 1, 2, ...</code>).</p>
<div id="input-is-a-data-frame-1" class="section level3">
<h3><span class="header-section-number">10.2.1</span> Input is a data frame</h3>
<p>Under <code>stack</code> mode, when input data is a single data frame containing one or more numeric columns, each numeric column defined in <code>numeric.column</code> will be treated as a single unit (recall that when <code>numeric.column</code> is not specified, all numeric columns are used). <code>ylim</code> is re-defined to <code>c(0.5, n+0.5)</code> in which <code>n</code> is number of numeric columns specified. <code>panel.fun</code> is applied iteratively on each numeric column and add graphics to the horizontal line <code>y = i</code>. In this case, actually <code>value</code> in e.g. <code>circos.genomicPoints()</code> doesn’t used for mapping the y positions, while replaced with <code>y = i</code> internally.</p>
<p>In each iteration, in <code>panel.fun</code>, <code>region</code> is still the genomic regions in current chromosome, but <code>value</code> only contains current numeric column plus all non-numeric columns. The value of the index of “current” numeric column can be obtained by <code>getI(...)</code>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">data =<span class="st"> </span><span class="kw">generateRandomBed</span>(<span class="dt">nc =</span> <span class="dv">2</span>)
<span class="kw">circos.genomicTrack</span>(data, <span class="dt">stack =</span> <span class="ot">TRUE</span>,
    <span class="dt">panel.fun =</span> <span class="cf">function</span>(region, value, ...) {
        i =<span class="st"> </span><span class="kw">getI</span>(...)
        <span class="kw">circos.genomicPoints</span>(region, value, <span class="dt">col =</span> i, ...)
})</code></pre></div>
</div>
<div id="input-is-a-list-of-data-frames-1" class="section level3">
<h3><span class="header-section-number">10.2.2</span> Input is a list of data frames</h3>
<p>When input data is a list of data frames, each data frame will be treated as a single unit. <code>ylim</code> is re-defined to <code>c(0.5, n+0.5)</code> in which <code>n</code> is the number of data frames. <code>panel.fun</code> will be applied iteratively on each data frame. In each iteration, in <code>panel.fun</code>, <code>region</code> is still the genomic regions in current chromosome, and <code>value</code> contains columns in current data frame excluding the first three columns. Graphics by low-level genomic functions will be added on the `horizontal’ bins.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">bed_list =<span class="st"> </span><span class="kw">list</span>(<span class="kw">generateRandomBed</span>(), <span class="kw">generateRandomBed</span>())
<span class="kw">circos.genomicTrack</span>(bed_list, <span class="dt">stack =</span> <span class="ot">TRUE</span>,
    <span class="dt">panel.fun =</span> <span class="cf">function</span>(region, value, ...) {
        i =<span class="st"> </span><span class="kw">getI</span>(...)
        <span class="kw">circos.genomicPoints</span>(region, value, ...)
})</code></pre></div>
<p>Under <code>stack</code> mode, if using a data frame with multiple numeric columns, graphics on all horizontal bins share the same genomic positions while if using a list of data frames, the genomic positions can be different.</p>
</div>
</div>
<div id="applications" class="section level2">
<h2><span class="header-section-number">10.3</span> Applications</h2>
<p>In this section, we will show several real examples of adding genomic graphics under different modes. Again, if you are not happy with these functionalities, you can simply re-implement your plot with the basic circlize functions.</p>
<div id="modes-points" class="section level3">
<h3><span class="header-section-number">10.3.1</span> Points</h3>
<p>To make plots more clear to look at, we only add graphics in the first quarter of the circle and initialize the plot only with chromosome 1.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">set.seed</span>(<span class="dv">999</span>)
<span class="kw">circos.par</span>(<span class="st">&quot;track.height&quot;</span> =<span class="st"> </span><span class="fl">0.1</span>, <span class="dt">start.degree =</span> <span class="dv">90</span>,
    <span class="dt">canvas.xlim =</span> <span class="kw">c</span>(<span class="dv">0</span>, <span class="dv">1</span>), <span class="dt">canvas.ylim =</span> <span class="kw">c</span>(<span class="dv">0</span>, <span class="dv">1</span>), <span class="dt">gap.degree =</span> <span class="dv">270</span>)
<span class="kw">circos.initializeWithIdeogram</span>(<span class="dt">chromosome.index =</span> <span class="st">&quot;chr1&quot;</span>, <span class="dt">plotType =</span> <span class="ot">NULL</span>)</code></pre></div>
<p>In the example figure (Figure <a href="modes-of-input.html#fig:genomic-application-points">10.1</a>) below, each track contains points under different modes.</p>
<p>In track A, it is the most normal way to add points. Here <code>bed</code> only contains one numeric column and points are added at the middle points of regions.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">bed =<span class="st"> </span><span class="kw">generateRandomBed</span>(<span class="dt">nr =</span> <span class="dv">300</span>)
<span class="kw">circos.genomicTrack</span>(bed, <span class="dt">panel.fun =</span> <span class="cf">function</span>(region, value, ...) {
    <span class="kw">circos.genomicPoints</span>(region, value, <span class="dt">pch =</span> <span class="dv">16</span>, <span class="dt">cex =</span> <span class="fl">0.5</span>, ...)
})</code></pre></div>
<p>In track B, if it is specified as <code>stack</code> mode, points are added in a horizontal line (or visually, a circular line).</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">circos.genomicTrack</span>(bed, <span class="dt">stack =</span> <span class="ot">TRUE</span>, 
    <span class="dt">panel.fun =</span> <span class="cf">function</span>(region, value, ...) {
        <span class="kw">circos.genomicPoints</span>(region, value, <span class="dt">pch =</span> <span class="dv">16</span>, <span class="dt">cex =</span> <span class="fl">0.5</span>,...)
        i =<span class="st"> </span><span class="kw">getI</span>(...)
        <span class="kw">circos.lines</span>(CELL_META<span class="op">$</span>cell.xlim, <span class="kw">c</span>(i, i), <span class="dt">lty =</span> <span class="dv">2</span>, <span class="dt">col =</span> <span class="st">&quot;#00000040&quot;</span>)
})</code></pre></div>
<p>In track C, the input data is a list of two data frames. <code>panel.fun</code> is applied iterately on each data frame. The index of “current” index can be obtained by <code>getI(...)</code>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">bed1 =<span class="st"> </span><span class="kw">generateRandomBed</span>(<span class="dt">nr =</span> <span class="dv">300</span>)
bed2 =<span class="st"> </span><span class="kw">generateRandomBed</span>(<span class="dt">nr =</span> <span class="dv">300</span>)
bed_list =<span class="st"> </span><span class="kw">list</span>(bed1, bed2)
<span class="kw">circos.genomicTrack</span>(bed_list, 
    <span class="dt">panel.fun =</span> <span class="cf">function</span>(region, value, ...) {
        i =<span class="st"> </span><span class="kw">getI</span>(...)
        <span class="kw">circos.genomicPoints</span>(region, value, <span class="dt">pch =</span> <span class="dv">16</span>, <span class="dt">cex =</span> <span class="fl">0.5</span>, <span class="dt">col =</span> i, ...)
})</code></pre></div>
<p>In track D, the list of data frames is plotted under <code>stack</code> mode. Graphics corresponding to each data frame are added to a horizontal line.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">circos.genomicTrack</span>(bed_list, <span class="dt">stack =</span> <span class="ot">TRUE</span>, 
    <span class="dt">panel.fun =</span> <span class="cf">function</span>(region, value, ...) {
        i =<span class="st"> </span><span class="kw">getI</span>(...)
        <span class="kw">circos.genomicPoints</span>(region, value, <span class="dt">pch =</span> <span class="dv">16</span>, <span class="dt">cex =</span> <span class="fl">0.5</span>, <span class="dt">col =</span> i, ...)
        <span class="kw">circos.lines</span>(CELL_META<span class="op">$</span>cell.xlim, <span class="kw">c</span>(i, i), <span class="dt">lty =</span> <span class="dv">2</span>, <span class="dt">col =</span> <span class="st">&quot;#00000040&quot;</span>)
})</code></pre></div>
<p>In track E, the data frame has four numeric columns. Under normal mode, all the four columns are used with the same genomic coordinates.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">bed =<span class="st"> </span><span class="kw">generateRandomBed</span>(<span class="dt">nr =</span> <span class="dv">300</span>, <span class="dt">nc =</span> <span class="dv">4</span>)
<span class="kw">circos.genomicTrack</span>(bed, 
    <span class="dt">panel.fun =</span> <span class="cf">function</span>(region, value, ...) {
        <span class="kw">circos.genomicPoints</span>(region, value, <span class="dt">pch =</span> <span class="dv">16</span>, <span class="dt">cex =</span> <span class="fl">0.5</span>, <span class="dt">col =</span> <span class="dv">1</span><span class="op">:</span><span class="dv">4</span>, ...)
})</code></pre></div>
<p>In track F, the data frame has four columns but is plotted under <code>stack</code> mode. Graphics for each column are added to a horizontal line. Current column can be obtained by <code>getI(...)</code>. Note here <code>value</code> in <code>panel.fun</code> is a data frame with only one column (which is the current numeric column).</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">bed =<span class="st"> </span><span class="kw">generateRandomBed</span>(<span class="dt">nr =</span> <span class="dv">300</span>, <span class="dt">nc =</span> <span class="dv">4</span>)
<span class="kw">circos.genomicTrack</span>(bed, <span class="dt">stack =</span> <span class="ot">TRUE</span>, 
    <span class="dt">panel.fun =</span> <span class="cf">function</span>(region, value, ...) {
        i =<span class="st"> </span><span class="kw">getI</span>(...)
        <span class="kw">circos.genomicPoints</span>(region, value, <span class="dt">pch =</span> <span class="dv">16</span>, <span class="dt">cex =</span> <span class="fl">0.5</span>, <span class="dt">col =</span> i, ...)
        <span class="kw">circos.lines</span>(CELL_META<span class="op">$</span>cell.xlim, <span class="kw">c</span>(i, i), <span class="dt">lty =</span> <span class="dv">2</span>, <span class="dt">col =</span> <span class="st">&quot;#00000040&quot;</span>)
})
<span class="kw">circos.clear</span>()</code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:genomic-application-points"></span>
<img src="10-modes-of-input_files/figure-html/genomic-application-points-1.svg" alt="Add points under different modes." width="384" />
<p class="caption">
Figure 10.1: Add points under different modes.
</p>
</div>
</div>
<div id="modes-lines" class="section level3">
<h3><span class="header-section-number">10.3.2</span> Lines</h3>
<p>Similar as previous figure, only the first quarter in the circle is visualized. Examples are shown in Figure <a href="modes-of-input.html#fig:genomic-application-lines">10.2</a>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">circos.par</span>(<span class="st">&quot;track.height&quot;</span> =<span class="st"> </span><span class="fl">0.08</span>, <span class="dt">start.degree =</span> <span class="dv">90</span>,
    <span class="dt">canvas.xlim =</span> <span class="kw">c</span>(<span class="dv">0</span>, <span class="dv">1</span>), <span class="dt">canvas.ylim =</span> <span class="kw">c</span>(<span class="dv">0</span>, <span class="dv">1</span>), <span class="dt">gap.degree =</span> <span class="dv">270</span>,
    <span class="dt">cell.padding =</span> <span class="kw">c</span>(<span class="dv">0</span>, <span class="dv">0</span>, <span class="dv">0</span>, <span class="dv">0</span>))
<span class="kw">circos.initializeWithIdeogram</span>(<span class="dt">chromosome.index =</span> <span class="st">&quot;chr1&quot;</span>, <span class="dt">plotType =</span> <span class="ot">NULL</span>)</code></pre></div>
<p>In track A, it is the most simple way to add lines. Middle points of regions are used as the values on x-axes.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">bed =<span class="st"> </span><span class="kw">generateRandomBed</span>(<span class="dt">nr =</span> <span class="dv">500</span>)
<span class="kw">circos.genomicTrack</span>(bed, 
    <span class="dt">panel.fun =</span> <span class="cf">function</span>(region, value, ...) {
        <span class="kw">circos.genomicLines</span>(region, value)
})</code></pre></div>
<p><code>circos.genomicLines()</code> is implemented by <code>circos.lines()</code>, thus, arguments supported in <code>circos.lines()</code> can also be in <code>circos.genomicLines()</code>. In track B, the area under the line is filled with color and in track C, type of the line is set to <code>h</code>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">circos.genomicTrack</span>(bed, 
    <span class="dt">panel.fun =</span> <span class="cf">function</span>(region, value, ...) {
        <span class="kw">circos.genomicLines</span>(region, value, <span class="dt">area =</span> <span class="ot">TRUE</span>)
})
<span class="kw">circos.genomicTrack</span>(bed, 
    <span class="dt">panel.fun =</span> <span class="cf">function</span>(region, value, ...) {
        <span class="kw">circos.genomicLines</span>(region, value, <span class="dt">type =</span> <span class="st">&quot;h&quot;</span>)
})</code></pre></div>
<p>In track D, the input is a list of data frames. <code>panel.fun</code> is applied to each data frame iterately.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">bed1 =<span class="st"> </span><span class="kw">generateRandomBed</span>(<span class="dt">nr =</span> <span class="dv">500</span>)
bed2 =<span class="st"> </span><span class="kw">generateRandomBed</span>(<span class="dt">nr =</span> <span class="dv">500</span>)
bed_list =<span class="st"> </span><span class="kw">list</span>(bed1, bed2)
<span class="kw">circos.genomicTrack</span>(bed_list, 
    <span class="dt">panel.fun =</span> <span class="cf">function</span>(region, value, ...) {
        i =<span class="st"> </span><span class="kw">getI</span>(...)
        <span class="kw">circos.genomicLines</span>(region, value, <span class="dt">col =</span> i, ...)
})</code></pre></div>
<p>In track E, the input is a list of data frames and is drawn under <code>stack</code> mode. Each genomic region is drawn as a horizontal segment and is put on a horizontal line where the width of the segment corresponds to the width of the genomc region. Under <code>stack</code> mode, for <code>circos.genomicLines()</code>, type of lines is only restricted to segments.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">circos.genomicTrack</span>(bed_list, <span class="dt">stack =</span> <span class="ot">TRUE</span>, 
    <span class="dt">panel.fun =</span> <span class="cf">function</span>(region, value, ...) {
        i =<span class="st"> </span><span class="kw">getI</span>(...)
        <span class="kw">circos.genomicLines</span>(region, value, <span class="dt">col =</span> i, ...)
})</code></pre></div>
<p>In track F, the input is a data frame with four numeric columns. Each column is drawn under the normal mode where the same genomic coordinates are shared.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">bed =<span class="st"> </span><span class="kw">generateRandomBed</span>(<span class="dt">nr =</span> <span class="dv">500</span>, <span class="dt">nc =</span> <span class="dv">4</span>)
<span class="kw">circos.genomicTrack</span>(bed, 
    <span class="dt">panel.fun =</span> <span class="cf">function</span>(region, value, ...) {
        <span class="kw">circos.genomicLines</span>(region, value, <span class="dt">col =</span> <span class="dv">1</span><span class="op">:</span><span class="dv">4</span>, ...)
})</code></pre></div>
<p>In track G, the data frame with four numeric columns are drawn under <code>stack</code> mode. All the four columns are drawn to four horizontal lines.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">bed =<span class="st"> </span><span class="kw">generateRandomBed</span>(<span class="dt">nr =</span> <span class="dv">500</span>, <span class="dt">nc =</span> <span class="dv">4</span>)
<span class="kw">circos.genomicTrack</span>(bed, <span class="dt">stack =</span> <span class="ot">TRUE</span>, 
    <span class="dt">panel.fun =</span> <span class="cf">function</span>(region, value, ...) {
        i =<span class="st"> </span><span class="kw">getI</span>(...)
        <span class="kw">circos.genomicLines</span>(region, value, <span class="dt">col =</span> i, ...)
})</code></pre></div>
<p>In track H, we specify <code>type</code> to <code>segment</code> and set different colors for segments. Note each segment is located at the y position defined in the numeric column.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">bed =<span class="st"> </span><span class="kw">generateRandomBed</span>(<span class="dt">nr =</span> <span class="dv">200</span>)
<span class="kw">circos.genomicTrack</span>(bed, 
    <span class="dt">panel.fun =</span> <span class="cf">function</span>(region, value, ...) {
        <span class="kw">circos.genomicLines</span>(region, value, <span class="dt">type =</span> <span class="st">&quot;segment&quot;</span>, <span class="dt">lwd =</span> <span class="dv">2</span>, 
            <span class="dt">col =</span> <span class="kw">rand_color</span>(<span class="kw">nrow</span>(region)), ...)
})
<span class="kw">circos.clear</span>()</code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:genomic-application-lines"></span>
<img src="10-modes-of-input_files/figure-html/genomic-application-lines-1.svg" alt="Add lines under different modes." width="384" />
<p class="caption">
Figure 10.2: Add lines under different modes.
</p>
</div>
</div>
<div id="modes-rectangles" class="section level3">
<h3><span class="header-section-number">10.3.3</span> Rectangles</h3>
<p>Again, only the first quarter of the circle is initialized. For rectangles, the filled colors are always used to represent numeric values. Here we define a color mapping function <code>col_fun</code> to map values to colors. Examples are in Figure <a href="modes-of-input.html#fig:genomic-application-rect">10.3</a>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">circos.par</span>(<span class="st">&quot;track.height&quot;</span> =<span class="st"> </span><span class="fl">0.15</span>, <span class="dt">start.degree =</span> <span class="dv">90</span>,
    <span class="dt">canvas.xlim =</span> <span class="kw">c</span>(<span class="dv">0</span>, <span class="dv">1</span>), <span class="dt">canvas.ylim =</span> <span class="kw">c</span>(<span class="dv">0</span>, <span class="dv">1</span>), <span class="dt">gap.degree =</span> <span class="dv">270</span>)
<span class="kw">circos.initializeWithIdeogram</span>(<span class="dt">chromosome.index =</span> <span class="st">&quot;chr1&quot;</span>, <span class="dt">plotType =</span> <span class="ot">NULL</span>)
col_fun =<span class="st"> </span><span class="kw">colorRamp2</span>(<span class="dt">breaks =</span> <span class="kw">c</span>(<span class="op">-</span><span class="dv">1</span>, <span class="dv">0</span>, <span class="dv">1</span>), <span class="dt">colors =</span> <span class="kw">c</span>(<span class="st">&quot;green&quot;</span>, <span class="st">&quot;black&quot;</span>, <span class="st">&quot;red&quot;</span>))</code></pre></div>
<p>To draw heatmaps, you probably want to use the <code>stack</code> mode. In track A, <code>bed</code> has four numeric columns and <code>stack</code> mode is used to arrange the heatmap. You can see rectangles are stacked for a certain genomic region.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">bed =<span class="st"> </span><span class="kw">generateRandomBed</span>(<span class="dt">nr =</span> <span class="dv">100</span>, <span class="dt">nc =</span> <span class="dv">4</span>)
<span class="kw">circos.genomicTrack</span>(bed, <span class="dt">stack =</span> <span class="ot">TRUE</span>, 
    <span class="dt">panel.fun =</span> <span class="cf">function</span>(region, value, ...) {
        <span class="kw">circos.genomicRect</span>(region, value, <span class="dt">col =</span> <span class="kw">col_fun</span>(value[[<span class="dv">1</span>]]), <span class="dt">border =</span> <span class="ot">NA</span>, ...)
})</code></pre></div>
<p>In track B, the input is a list of data frames. Under <code>stack</code> mode, each data frame is added to a horizontal line. Since genomic positions for different data frames can be different, you may see in the figure, positions for the two sets of rectangles are different.</p>
<p>Under <code>stack</code> mode, by default, the height of rectangles is internally set to make them completely fill the cell in the vertical direction. <code>ytop</code> and <code>ybottom</code> can be used to adjust the height of rectangles. Note each line of rectangles is at <code>y = i</code> and the default height of rectangles are 1.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">bed1 =<span class="st"> </span><span class="kw">generateRandomBed</span>(<span class="dt">nr =</span> <span class="dv">100</span>)
bed2 =<span class="st"> </span><span class="kw">generateRandomBed</span>(<span class="dt">nr =</span> <span class="dv">100</span>)
bed_list =<span class="st"> </span><span class="kw">list</span>(bed1, bed2)
<span class="kw">circos.genomicTrack</span>(bed_list, <span class="dt">stack =</span> <span class="ot">TRUE</span>, 
    <span class="dt">panel.fun =</span> <span class="cf">function</span>(region, value, ...) {
        i =<span class="st"> </span><span class="kw">getI</span>(...)
        <span class="kw">circos.genomicRect</span>(region, value, <span class="dt">ytop =</span> i <span class="op">+</span><span class="st"> </span><span class="fl">0.3</span>, <span class="dt">ybottom =</span> i <span class="op">-</span><span class="st"> </span><span class="fl">0.3</span>,
            <span class="dt">col =</span> <span class="kw">col_fun</span>(value[[<span class="dv">1</span>]]), ...)
})</code></pre></div>
<p>In track C, we implement same graphics as in track B, but with the normal mode. Under <code>stack</code> mode, data range on y axes and positions of rectangles are adjusted internally. Here we explicitly adjust it under the normal mode.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">circos.genomicTrack</span>(bed_list, <span class="dt">ylim =</span> <span class="kw">c</span>(<span class="fl">0.5</span>, <span class="fl">2.5</span>), 
    <span class="dt">panel.fun =</span> <span class="cf">function</span>(region, value, ...) {
        i =<span class="st"> </span><span class="kw">getI</span>(...)
        <span class="kw">circos.genomicRect</span>(region, value, <span class="dt">ytop =</span> i <span class="op">+</span><span class="st"> </span><span class="fl">0.3</span>, <span class="dt">ybottom =</span> i <span class="op">-</span><span class="st"> </span><span class="fl">0.3</span>, 
            <span class="dt">col =</span> <span class="kw">col_fun</span>(value[[<span class="dv">1</span>]]), ...)
})</code></pre></div>
<p>In track D, rectangles are used to make barplots. We specify the position of the top of bars by <code>ytop.column</code> (1 means the first column in <code>value</code>).</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">bed =<span class="st"> </span><span class="kw">generateRandomBed</span>(<span class="dt">nr =</span> <span class="dv">200</span>)
<span class="kw">circos.genomicTrack</span>(bed, 
    <span class="dt">panel.fun =</span> <span class="cf">function</span>(region, value, ...) {
        <span class="kw">circos.genomicRect</span>(region, value, <span class="dt">ytop.column =</span> <span class="dv">1</span>, <span class="dt">ybottom =</span> <span class="dv">0</span>, 
            <span class="dt">col =</span> <span class="kw">ifelse</span>(value[[<span class="dv">1</span>]] <span class="op">&gt;</span><span class="st"> </span><span class="dv">0</span>, <span class="st">&quot;red&quot;</span>, <span class="st">&quot;green&quot;</span>), ...)
        <span class="kw">circos.lines</span>(CELL_META<span class="op">$</span>cell.xlim, <span class="kw">c</span>(<span class="dv">0</span>, <span class="dv">0</span>), <span class="dt">lty =</span> <span class="dv">2</span>, <span class="dt">col =</span> <span class="st">&quot;#00000040&quot;</span>)
})
<span class="kw">circos.clear</span>()</code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:genomic-application-rect"></span>
<img src="10-modes-of-input_files/figure-html/genomic-application-rect-1.svg" alt="Add rectangles under different modes." width="384" />
<p class="caption">
Figure 10.3: Add rectangles under different modes.
</p>
</div>

</div>
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="genomic-plotting-region.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="high-level-genomic-functions.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"google": false,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"all": ["facebook", "google", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/jokergoo/circlize_book/edit/master/10-modes-of-input.Rmd",
"text": "Edit"
},
"history": {
"link": null,
"text": null
},
"download": ["circlize-book.pdf", "circlize-book.epub"],
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:" && /^https?:/.test(src))
      src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
